state-chain/chains/src/sol/api.rs

```diff
@@ -60,35 +71,56 @@ pub trait SolanaEnvironment:
 					SolanaTransactionBuildingError::CannotLookupVaultProgramDataAccount,
 				SolanaEnvAccountLookupKey::UpgradeManagerProgramDataAccount =>
 					SolanaTransactionBuildingError::CannotLookupUpgradeManagerProgramDataAccount,
-				SolanaEnvAccountLookupKey::TokenMintPubkey =>
-					SolanaTransactionBuildingError::CannotLookupTokenMintPubkey,
-				SolanaEnvAccountLookupKey::TokenVaultAssociatedTokenAccount =>
-					SolanaTransactionBuildingError::CannotLookupTokenVaultAssociatedTokenAccount,
-				SolanaEnvAccountLookupKey::TokenVaultPdaAccount =>
-					SolanaTransactionBuildingError::CannotLookupTokenVaultPdaAccount,
 			},
 		)
 	}
 
+	fn lookup_account_retain(
+		key: SolanaEnvAccountLookupKey,
+		sol_environment: &mut HashMap<SolanaEnvAccountLookupKey, SolAddress>,
```

Comment:
Maybe this hasn't been linked with the runtime yet, but i'm pretty sure HashMaps are not supported. Prefer to use BTreeMap.

Reply:
Also, not quite sure what the purpose of this is - I guess some sort of cache? If so then I would suggest making it an implementation detail rather than baking it into the trait definition.

But in fact I would suggest just removing this - I think substrate already caches storage reads. And there is a separate issue to change the environment lookups. If necessary we can tackle any optimisations then.

Reply:
I wasn't sure that substrate caches storage reads, I thought it doesnt. If we are sure that is the case then yeah it is not needed. In case it doesnt, then I think it makes sense to do this since 10 fetches would mean 10 storage reads for example.

Reply:
If we read the whole environment in one go (as a single struct), instead of storing all the addresses separately, then we don't need caching. 

We can just get the environment and then we have all the addresses we need already available.

Reply:
Yeah I agree, I also think the caching design might be a bit over-complicated here. Since we only do 1 egress per block, by getting all the environment in the AllBatch layer, we are already doing only 1 storage read per block. 
By hiding the caching logic in the environment implementation, it is not obvious that caching is taking place, and it looks like we are doing repeated reads per environment variable per Transfer, which is not the case. I tend to think that it would be better to make it immediately clear that we are retrieving the environments _once_ per AllBatch (hence my original design of getting all the variables, then loop over each Transfer).

Reply:
Regarding using HashMaps: I am not sure if Hashmaps are supported by runtime or not, but happy to make it BTreeMap anyways.

Regarding reading the whole env at once: Ok, in the interest of simplicity i will see how to lump together the different environment elements into 1, remove caching and minimize reads.

Reply:
> Regarding using HashMaps: I am not sure if Hashmaps are supported by runtime or not, but happy to make it BTreeMap anyways.

HashMap is imported from std (`use std::collections::HashMap;`), which is not available in the runtime. 

I don't think we need to lump *everything* into one. I can also see why caching *might* make sense here once we support multiple assets. But either way, this is not the way to implement the cache: it should be invisible to the trait methods. For example, it should be possible to implement the environment *without* a cache (for example in the mocks).

Reply:
> HashMap is imported from std (`use std::collections::HashMap;`), which is not available in the runtime.

Oh yes, didn't see that. Thats what happens when you use vs code's auto-import :)

> I don't think we need to lump _everything_ into one. I can also see why caching _might_ make sense here once we support multiple assets. But either way, this is not the way to implement the cache: it should be invisible to the trait methods. For example, it should be possible to implement the environment _without_ a cache (for example in the mocks).

I don't think I agree with this. Its a matter of perspective on what the SolanaEnvironment trait is. I don't see why a general Environment trait cant have two different functions, one a normal lookup function and the another that lookup while caching in a provided argument.  Sometimes, you might wanna cache, sometimes you might not want to cache. We could have another trait specifically for caching that makes it explicit but that would be just splitting the trait. The caching cant be done inside the ```lookup``` functions since we need to pass the caching variable from the apicall construction inwards into the lookup which we cant do with the current ```ChainEnvironment``` trait.

Moreover, the confusion for me comes from the fact that i think the SolanaEnvironment trait is unnecessary since 1) its a single use trait and 2) it implements all the functions inside by default. We could have all the trait functions as normal functions without and outside the trait. This is what I did at first when I implemented the caching but then I realized that the function is similar to the other functions inside SolanaEnvironment trait and to keep things similar I put it inside.